src/modules/roc_core/target_posix/roc_core/colors.cpp

```diff
@@ -0,0 +1,63 @@
+/*
+ * Copyright (c) 2019 Roc authors
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+
+#include <stdio.h>
+#include <string.h>
+#include <unistd.h>
+
+#include "roc_core/colors.h"
+
+namespace roc {
+namespace core {
+
+//! ANSI Color Codes.
+#define COLOR_NONE ""
+#define COLOR_START "\033["
+#define COLOR_END "m"
+#define COLOR_RESET COLOR_START "0" COLOR_END
+#define COLOR_SEPARATOR ";"
+#define COLOR_BOLD "1"
+#define COLOR_RED "31"
+#define COLOR_BLUE "34"
+
+namespace {
+char const* colors_to_levels(LogLevel level) {
+    char const* color = COLOR_NONE;
+    switch (level) {
+    case LogNone: //!< No color for Loglevel None.
+        break;
+    case LogError: //!< Bold Red for Loglevel Error.
+        color = COLOR_START COLOR_BOLD COLOR_SEPARATOR COLOR_RED COLOR_END;
+        break;
+    case LogInfo: //!< Bold Blue for Loglevel Info.
+        color = COLOR_START COLOR_BOLD COLOR_SEPARATOR COLOR_BLUE COLOR_END;
+        break;
+    case LogDebug: //!< No color for Loglevel Debug.
+        break;
+    case LogTrace: //!< No color for Loglevel Trace.
+        break;
+    }
+    return color;
+}
+} // namespace
+
+bool colors_available() {
+    char* term = getenv("TERM");
+    return isatty(STDERR_FILENO) && term && strncmp("dumb", term, 4) != 0;
```

Comment:
Could you explain why do we need to check `TERM`?

There are two concerns here:

* We're performing a search in environment for every log message. Not very nice.

* What is worse, our library is now calling getenv() from its threads (when it logs something). If the user code outside of our library will call setenv() from another thread, we'll get a race and probably a crash.

I'd prefer to avoid calling getenv() if it's not necessary. If it's necessary, I'd prefer to call it only in tools and only once. E.g. we could add a function colors_init() and call it from main().

Reply:
Agree with your points :)
The intention behind $TERM check is actually about convention – many POSIX command line tools supports no-colorized outputs when $TERM=dumb; e.g. coreutils and git.
I firstly thought about adding static variables to the `colors_available()`, but as you suggested, calling it only from tools makes more sense. Thanks!

Reply:
Ah, I see. If it's a common convention, it would not be very good to ignore it. Maybe for now it would be better to disable coloring at all when roc is used as a library.

Otherwise we should provide a C API to forcibly enable and disable colors and probably to process the environment in a thread-safe manner, e.g. some global initialization function that is allowed to call getenv(). This looks like an overkill for this feature.

BTW, current API allows to provide a custom log handler, so that the user can implement colored output on top of roc, if needed.

I think we can replace ColorsAlways/ColorsNever/ColorsAuto with just two modes ColorsEnabled/ColorsDisabled (or with a boolean) and by default use ColorsDisabled. Command-line will continue to use always|never|auto and will set ColorsEnabled/ColorsDisabled depending on the command-line option and colors_available() function.